### PR TITLE
Edge 1005 async inbound

### PR DESCRIPTION
**Pull request by smckinnon-bw**
A fairly significant upgrade to the bridge sample application
- made the setup of the phone leg and the sip interconnect leg decoupled and asynchronous
- added inbound calling to the designated phone
- extended the readme
- changed the web front-end to use webHooks to support incoming call notifications

### CODE REVIEW

File: frontend/src/App.tsx [hunk 1]
```diff
@@ -14,16 +17,54 @@ const App: React.FC = () => {
...
+   */
+
+  interface CallState {
+    event: string; // registered, callStateUpdate
+    token?: string;
+    tn?: string;
+    callState?: string;
+  }
+
+  interface clientEvent {
```
Comment (carrington-bw) [inline frontend/src/App.tsx:39]: Make the C capitalized

File: frontend/src/App.tsx [hunk 2]
```diff
@@ -96,7 +131,10 @@ const App: React.FC = () => {
   return (
     <div className="App">
       <header className="App-header">
-        <div>WebRTC Hello World</div>
+        <div>WebRTC Voice Calls - using asynchronous Bridge</div>
```
Comment (carrington-bw) [inline frontend/src/App.tsx:134]: Capitalize U and A... It looks more consistent

Comment (carrington-bw) [reply frontend/src/App.tsx:134]: Or make B lowercase

File: frontend/src/App.tsx [hunk 3]
```diff
@@ -115,40 +153,41 @@ const App: React.FC = () => {
...
+                  type="text"
+                  name="numberToDial"
+                  id="numberToDial"
+                  placeholder="enter a phone number"
+                  style={{ height: "30px", marginLeft: "10px" }}
+                  onChange={updateTn}
+                />
+              </React.Fragment>
+            ) : (
+              <span>. . . . . .</span>
```
Comment (carrington-bw) [inline frontend/src/App.tsx:186]: The Action:...... looked a little weird in the browser. I don't have a specific suggestion but I would recommend something different

File: src/server.ts [hunk 4]
```diff
@@ -37,10 +41,10 @@ const voiceApplicationPhoneNumber = <string>process.env.BW_NUMBER; // the 'from'
...
 const voiceCallbackUrl = <string>process.env.BASE_CALLBACK_URL;
 
-console.log(
-  "call control url is...",
-  process.env.BANDWIDTH_WEBRTC_CALL_CONTROL_URL
-);
+// console.log(
+//   "call control url is...",
+//   process.env.BANDWIDTH_WEBRTC_CALL_CONTROL_URL
+// );
```
Comment (carrington-bw) [inline src/server.ts:47]: the callControlUrl is never used

File: src/server.ts [hunk 5]
```diff
@@ -119,77 +144,86 @@ app.post("/killConnection", async (req, res) => {
     webParticipant &&
     req.body.participantId == webParticipant.id
   ) {
-    // TODO - move the null check inside of the functions
     console.log("deallocating all configured resources on exit");
     await killSipUriLeg(bridgeParticipant);
     await deleteParticipant(bridgeParticipant);
     await deleteParticipant(webParticipant);
     await deleteSession();
+    clientWs.close();
```
Comment (carrington-bw) [inline src/server.ts:152]: You might want to put an existence check around this:

if (clientWs) {
   clientWs.close()
}

File: src/server.ts [hunk 6]
```diff
@@ -119,77 +144,86 @@ app.post("/killConnection", async (req, res) => {
...
- */
-app.post("/bridgeCallAnswered", async (req, res) => {
-  const callId = req.body.callId;
-  console.log(
-    `received answered callback for bridging call ${callId} to ${req.body.to}`
-  );
+  // setup the interconnection to webRTC
+  bridgeParticipant = await createParticipant("hello-world-phone");
+  // TODO - confirm the async behavior on call setup
+  callSipUri(bridgeParticipant);
```
Comment (carrington-bw) [inline src/server.ts:168]: callSipUri is async... You might want to put an await here

File: src/server.ts [hunk 7]
```diff
@@ -201,92 +235,128 @@ app.post("/callAnswered", async (req, res) => {
...
-  var response = new Response();
   response.add(speakSentence);
-  response.add(bridge);
 
-  let myResp = await response.toBxml();
-  console.log(myResp);
+  const bridgecallId: string = findCall("bridge");
+  if (bridgecallId) {
+    // the sipinterconnect call exists - bridge them
+    var bridge = new Bridge({
```
Comment (carrington-bw) [inline src/server.ts:253]: Global change to make... Try to replace var with either const or let. Usage of var is discouraged

File: src/server.ts [hunk 8]
```diff
@@ -201,92 +235,128 @@ app.post("/callAnswered", async (req, res) => {
...
+  // if there is another call present, around bridge it in, otherwise wait for another call.
+  let otherCallId = findCall("outbound") || findCall("inbound"); // look for any calls that are hanging around.
+  if (otherCallId) {
+    // there is an existing voice call that should be bridged
+    console.log(
+      "bridge the preexisting voice call: ",
+      otherCallId,
+      voiceCalls.get(otherCallId)
+    );
+    // the sipinterconnect call exists - bridge them
```
Comment (carrington-bw) [inline src/server.ts:307]: An external user isn't going to know what sipinterconnect is

File: src/server.ts [hunk 9]
```diff
@@ -201,92 +235,128 @@ app.post("/callAnswered", async (req, res) => {
...
-});
 
-app.post("/bridgeTargetComplete", async (req, res) => {
-  console.log("bridgeTargetComplete update: ", req.body);
+  let myResp = await response.toBxml();
+  res.send(myResp);
+  updateCallStatus("connected");
+});
 
+/**
```
Comment (carrington-bw) [inline src/server.ts:326]: Thanks for commenting all of the endpoints... It's very helpful

File: src/server.ts [hunk 10]
```diff
@@ -201,92 +235,128 @@ app.post("/callAnswered", async (req, res) => {
...
+      console.log(`received endBridgeLeg event for call ${callId}`);
 
-      var speakSentence = new SpeakSentence({
-        sentence: "Make another call and wait until that call answers",
-      });
+      // remove the SIP interconnect
+      await killSipUriLeg(bridgeParticipant);
+      await deleteParticipant(bridgeParticipant);
+
+      // var speakSentence = new SpeakSentence({
```
Comment (carrington-bw) [inline src/server.ts:344]: Maybe make a pass at cleaning up commented code

File: frontend/src/App.tsx (hunk 1 above)
Comment (smckinnon-bw) [reply frontend/src/App.tsx:39]: resolved

File: frontend/src/App.tsx (hunk 2 above)
Comment (smckinnon-bw) [reply frontend/src/App.tsx:134]: b lowercase.

File: frontend/src/App.tsx (hunk 3 above)
Comment (smckinnon-bw) [reply frontend/src/App.tsx:186]: Agreed - I'll have to mess with it.

File: src/server.ts (hunk 4 above)
Comment (smckinnon-bw) [reply src/server.ts:47]: thanks

File: src/server.ts [hunk 11]
```diff
@@ -59,7 +63,19 @@ interface ParticipantInfo {
 interface CallData {
   from: string;
   to: string;
-  bridge: boolean;
+  callType: string;
+}
+
+interface callState {
```
[the remark this answers is not in the document]
Comment (smckinnon-bw) [reply src/server.ts:69]: done

File: src/server.ts (hunk 5 above)
Comment (smckinnon-bw) [reply src/server.ts:152]: thanks

File: src/server.ts (hunk 6 above)
Comment (smckinnon-bw) [reply src/server.ts:168]: that was intentionally async, but I put a comment in to ensure that was clear.

File: src/server.ts (hunk 7 above)
Comment (smckinnon-bw) [reply src/server.ts:253]: bad Steve :-)   I'm going to blame my bad habit of clone and modify :-(  Thanks for the reminder.

File: src/server.ts (hunk 8 above)
Comment (smckinnon-bw) [reply src/server.ts:307]: thanks

File: src/server.ts (hunk 9 above)
Comment (smckinnon-bw) [reply src/server.ts:326]: :+1:

File: src/server.ts (hunk 10 above)
Comment (smckinnon-bw) [reply src/server.ts:344]: I will do so.

File: frontend/src/App.tsx (hunk 3 above)
Comment (smckinnon-bw) [reply frontend/src/App.tsx:186]: cleaned up a bit - no more 'Action' if there is no action that is possible.  Still room for explicit answer and hangup if we get around to it, but fearing that it is already too complex.

**Review by carrington-bw — APPROVED** [review]

**Pull request merged**